### PR TITLE
DSR-1942 Added object data structures and methods

### PR DESCRIPTION
**Pull request by dpazj**

### CODE REVIEW

File: src/characteristic.c [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+ * IoTech Ltd
+ *
+ **********************************************************************/
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+
+#include "characteristic.h"
+
+characteristic_t *characteristic_new (const char *UUID, const char *service_path, descriptor_t *descriptors)
```
Comment (Tylerg2471) [inline src/characteristic.c]: UUID should be lower case (coding standards)

File: src/characteristic.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+#define BLE_SIM_CHARACTERISTIC_H
+
+#include <stdint.h>
+
+#include "defines.h"
+#include "descriptor.h"
+
+typedef struct characteristic_t
+{
+  char *UUID; //128-bit characteristic UUID.
```
Comment (Tylerg2471) [inline src/characteristic.h]: UUID should be lower case (coding standards)

File: src/characteristic.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  uint8_t * value;
+  uint32_t value_size;
+  bool notifying;
+  uint32_t flags;
+  descriptor_t *descriptors;
+
+  struct characteristic_t *next;
+} characteristic_t;
+
+characteristic_t *characteristic_new (const char *UUID, const char *service_path, descriptor_t *descriptors);
```
Comment (Tylerg2471) [inline src/characteristic.h]: Doxygen comments should be added for all functions

File: src/characteristic.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+typedef struct characteristic_t
+{
+  char *UUID; //128-bit characteristic UUID.
+  char *service_path; //Object path of the GATT service the characteristic belongs to.
+  uint8_t * value;
+  uint32_t value_size;
+  bool notifying;
+  uint32_t flags;
+  descriptor_t *descriptors;
+
```
Comment (Tylerg2471) [inline src/characteristic.h]: Newline can be removed

File: src/dbusutils.c [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  DBusError err;
+  dbus_error_init (&err);
+
+  DBusMessage *dbus_msg = dbus_message_new_method_call (bus_name, path, iface, method);
+  if (dbus_msg == NULL)
+  {
+    return false;
+  }
+
+  DBusMessage *dbus_reply = dbus_connection_send_with_reply_and_block (connection, dbus_msg, 1000, &err);
```
Comment (Tylerg2471) [inline src/dbusutils.c]: Magic number '1000' should be defined

File: src/dbusutils.c [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    dbus_error_free (&err);
+    return false;
+  }
+
+  dbus_message_unref (dbus_reply);
+  dbus_error_free (&err);
+  return true;
+}
+
+
```
Comment (Tylerg2471) [inline src/dbusutils.c]: newline can be removed

File: src/dbusutils.h [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+#ifndef BLE_SIM_DBUSUTILS_H
+#define BLE_SIM_DBUSUTILS_H
+
+#include <stdbool.h>
+#include <dbus/dbus.h>
+
+#include "defines.h"
+
+DBusConnection * dbusutils_get_connection (void);
+bool dbusutils_request_application_bus_name (DBusConnection *connection);
```
Comment (Tylerg2471) [inline src/dbusutils.h:28]: Doxygen comments should be added for all functions

File: src/defines.h [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * Copyright (c) 2021
+ * IoTech Ltd
+ *
+ **********************************************************************/
+
+#ifndef BLE_SIM_DEFINES_H
+#define BLE_SIM_DEFINES_H
+
+
```
Comment (Tylerg2471) [inline src/defines.h]: Newline can be removed

File: src/descriptor.c [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+ * IoTech Ltd
+ *
+ **********************************************************************/
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "descriptor.h"
+
+descriptor_t *descriptor_new (const char *UUID, const char *characteristic_path)
```
Comment (Tylerg2471) [inline src/descriptor.c]: UUID should be lower case (coding standards)

File: src/descriptor.h [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+ **********************************************************************/
+
+#ifndef BLE_SIM_DESCRIPTOR_H
+#define BLE_SIM_DESCRIPTOR_H
+
+#include <stdint.h>
+
+typedef struct descriptor_t
+{
+  char *UUID; //128-bit descriptor UUID.
```
Comment (Tylerg2471) [inline src/descriptor.h]: UUID should be lower case (coding standards)

File: src/descriptor.h [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  char *characteristic_path; //Object path of the GATT characteristic the descriptor belongs to.
+  uint8_t * value;
+  uint32_t value_size;
+  uint16_t flags;
+
```
Comment (Tylerg2471) [inline src/descriptor.h]: newline can be removed

File: src/descriptor.h [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+  struct descriptor_t *next;
+} descriptor_t;
+
+descriptor_t *descriptor_new (const char *UUID, const char *characteristic_path);
```
Comment (Tylerg2471) [inline src/descriptor.h]: Doxygen comments should be added for all functions

File: src/device.h [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+ **********************************************************************/
+
+#ifndef BLE_SIM_DEVICE_H
+#define BLE_SIM_DEVICE_H
+
+#include <stdlib.h>
+
+#include "defines.h"
+#include "server.h"
+
```
Comment (Tylerg2471) [inline src/device.h:15]: Newline can be removed

File: src/device.h [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+typedef struct device_t
+{
+  server_t *gatt_server;
+  char *controller; //does nothing for now - in future this will be the controller this "device" is exposed on
+  char *device_name;
+  struct device_t *next;
+
```
Comment (Tylerg2471) [inline src/device.h]: newline can be removed

File: src/device.h [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+} device_t;
+
+device_t * device_get_device (const char* device_name);
```
Comment (Tylerg2471) [inline src/device.h:30]: Doxygen comments should be added for all functions

File: src/main.c [hunk 16]
```diff
@@ -1,27 +1,99 @@
...
+#include "service.h"
+#include "characteristic.h"
+#include "descriptor.h"
 
-int main (int argc, char *argv[]) 
+DBusConnection* global_dbus_connection;
+
+static DBusHandlerResult filter_message (DBusConnection *connection, DBusMessage *message, void *data)
+{
+
```
Comment (Tylerg2471) [inline src/main.c]: newline can be removed.

File: src/main.c [hunk 17]
```diff
@@ -1,27 +1,99 @@
...
+          dbus_message_get_destination (message),
+          dbus_message_get_path (message),
+          dbus_message_get_interface (message),
+          dbus_message_get_member (message),
+          dbus_message_get_type (message) == DBUS_MESSAGE_TYPE_ERROR ?
+          dbus_message_get_error_name (message) : ""
+        );
+
+  // if (dbus_message_is_method_call(msg, "test.method.Type", "Method"))
+  // if (dbus_message_is_signal(msg, "test.signal.Type", "Test"))
```
Comment (Tylerg2471) [inline src/main.c]: Question: Why are these commented out? If its going to be used later could add in a TODO

File: src/main.c [hunk 18]
```diff
@@ -1,27 +1,99 @@
...
-  if (NULL == conn) { 
-    return 1; 
+
+  dbus_connection_flush (global_dbus_connection);
+  dbus_connection_unref (global_dbus_connection);
+}
+
+int main (int argc, char *argv[]) 
+{
+
```
Comment (Tylerg2471) [inline src/main.c]: newline can be removed

File: src/main.c [hunk 19]
```diff
@@ -1,27 +1,99 @@
...
+  {
+    return 1;
   }
-  dbus_connection_close (conn);
+
+  const char *devname = "test-dev";
+
+  device_add (devname);
+
+  device_add_service (devname, service_new("0000180d-0000-1000-8000-00805f9b34fb", "/path/to/dev", true, NULL) );
```
Comment (Tylerg2471) [inline src/main.c]: String literals should be defined.

File: src/server.c [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "service.h"
+#include "server.h"
+#include "dbusutils.h"
+
+server_t *server_new (const char *object_path, service_t *services)
+{
+
```
Comment (Tylerg2471) [inline src/server.c]: newline can be removed

File: src/server.h [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+#include <dbus/dbus.h>
+
+#include "defines.h"
+#include "service.h"
+
+typedef struct server_t
+{
+  char * object_path; //dbus object path
+  service_t *services;
+
```
Comment (Tylerg2471) [inline src/server.h]: newline can be removed

File: src/server.h [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+#include "service.h"
+
+typedef struct server_t
+{
+  char * object_path; //dbus object path
+  service_t *services;
+  
+} server_t;
+
+server_t *server_new (const char *object_path, service_t *services);
```
Comment (Tylerg2471) [inline src/server.h:24]: Doxygen comments should be added for all functions

File: src/service.c [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+ * Copyright (c) 2021
+ * IoTech Ltd
+ *
+ **********************************************************************/
+#include <stdio.h>
+#include <string.h>
+
+#include "service.h"
+
+service_t *service_new (const char* UUID, const char *device_path, bool primary, characteristic_t *characteristics)
```
Comment (Tylerg2471) [inline src/service.c]: UUID should be lower case (coding standards)

File: src/service.h [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+#include <stdint.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "defines.h"
+#include "characteristic.h"
+
+typedef struct service_t
+{
+  char *UUID; //128-bit service UUID
```
Comment (Tylerg2471) [inline src/service.h]: UUID should be lower case (coding standards)

File: src/service.h [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+  char *device_path; // Object path of the Bluetooth device the service belongs to
+  bool primary;
+  characteristic_t * characteristics;
+
```
Comment (Tylerg2471) [inline src/service.h]: newline can be removed

File: src/service.h [hunk 26]
```diff
@@ -0,0 +1,39 @@
...
+{
+  char *UUID; //128-bit service UUID
+  char *device_path; // Object path of the Bluetooth device the service belongs to
+  bool primary;
+  characteristic_t * characteristics;
+  
+  struct service_t *next;
+} service_t;
+
+service_t *service_new (const char* UUID, const char *device_path, bool primary, characteristic_t *characteristics);
```
Comment (Tylerg2471) [inline src/service.h]: Doxygen comments should be added for all functions

File: src/service.h [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+void service_free (service_t *service);
+
+characteristic_t *service_get_characteristic (service_t *service, const char* characteristic_uuid);
+bool service_add_characteristic (service_t *service, characteristic_t *characteristic);
+
+
```
Comment (Tylerg2471) [inline src/service.h:63]: newline can be removed

**Review by Tylerg2471 — CHANGES_REQUESTED** [review]
Few comments to address

File: src/server.h (hunk 22 above)
Comment (dpazj) [reply src/server.h:24]: This file has been removed in the current branch I am working on so little point in adding doxygen

File: src/characteristic.h [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+ * @param descriptors characteristic descriptors 
+ * @return initialised characteristic  
+ **/
+characteristic_t *characteristic_new (const char *uuid, const char *service_path, descriptor_t *descriptors);
+
+/**
+ * Frees a characteristic_t and it's values
+ * 
+ * @param characteristic characteristic to free 
+ * @return void
```
Comment (Tylerg2471) [inline src/characteristic.h]: void functions don't return anything so doxygen shouldn't have a return parameter

File: src/characteristic.h [hunk 29]
```diff
@@ -0,0 +1,76 @@
...
+ * @return found descriptor or NULL if not found
+ **/
+descriptor_t *characteristic_get_descriptor (characteristic_t *characteristic, const char *descriptor_uuid);
+
+/**
+ * Adds a descriptor to the characteristic
+ * 
+ * @param characteristic characteristic to add the descriptor to
+ * @param descriptor descriptor to add 
+ * @return found descriptor or NULL if not found
```
Comment (Tylerg2471) [inline src/characteristic.h]: return is incorrect as it is bool and not a pointer it will either be true or false

File: src/main.c [hunk 30]
```diff
@@ -1,27 +1,104 @@
...
+#define TST_SRVC2 "0000180d-0000-1000-8000-00805f9b34f1"
+#define TST_SRVC3 "0000180d-0000-1000-8000-00805f9b34f2"
+#define TST_SRVC4 "0000180d-0000-1000-8000-00805f9b34f3"
+
+#define TST_CHR1 "00002a38-0000-1000-8000-00805f9b34fb"
+#define TST_CHR2 "00002a39-0000-1000-8000-00805f9b34fb"
+
+#define TST_DESC1 "12345678-1234-5678-1234-56789abcdef2"
+
+
```
Comment (Tylerg2471) [inline src/main.c]: newline can be removed

File: src/main.c [hunk 31]
```diff
@@ -1,27 +1,104 @@
...
+  {
+    return 1;
   }
-  dbus_connection_close (conn);
+
+  const char *devname = "test-dev";
+
+  device_add (devname);
+
+  device_add_service (devname, service_new(TST_SRVC1, "/path/to/dev", true, NULL) );
```
Comment (Tylerg2471) [inline src/main.c]: String literal can be defined

File: src/main.c [hunk 32]
```diff
@@ -1,27 +1,104 @@
...
+  device_add_service (devname, service_new(TST_SRVC2, "/path/to/dev", true, NULL) );
+  device_add_service (devname, service_new(TST_SRVC3, "/path/to/dev", true, NULL) );
+  device_add_service (devname, service_new(TST_SRVC4, "/path/to/dev", true, NULL) );
+
+  device_add_characteristic (devname, TST_SRVC1, characteristic_new ( TST_CHR1, "path", NULL));
```
Comment (Tylerg2471) [inline src/main.c]: String literal can be defined

File: src/service.h [hunk 33]
```diff
@@ -0,0 +1,68 @@
...
+ * @return found characteristic or NULL if not found
+ **/
+characteristic_t *service_get_characteristic (service_t *service, const char* characteristic_uuid);
+
+/**
+ * Adds a characteristic to the service
+ * 
+ * @param service service to add the characteristic to
+ * @param characteristic characteristic to add 
+ * @return found descriptor or NULL if not found
```
Comment (Tylerg2471) [inline src/service.h]: return doxygen comment is incorrect.

**Review by Tylerg2471 — CHANGES_REQUESTED** [review]
Some comments to look at

**Review by Tylerg2471 — APPROVED** [review]
LGTM

**Pull request merged**